File: mobile/modules/orca-mobile-web-shell/android/src/main/java/expo/modules/orcamobilewebshell/MobileWebShellResponseHeaders.kt
package expo.modules.orcamobilewebshell

/**
 * The headers one served asset answers with. Content-Type is not among them: `WebResourceResponse`
 * takes the mime type and the encoding as separate arguments.
 *
 * The policy header rides the document and nothing else: on a script or a stylesheet response it is
 * inert, and sending it everywhere would hide which response is the one that has to carry it.
 */
internal fun mobileWebShellResponseHeaders(path: String, byteCount: Int): Map<String, String> {
  val headers = mutableMapOf(
    "Content-Length" to byteCount.toString(),
    "Cache-Control" to "no-store",
    "X-Content-Type-Options" to "nosniff"
  )
  if (path == "/") {
    headers["Content-Security-Policy"] = MOBILE_WEB_SHELL_CSP
    // The document's origin here is a hashed host,
    // `https://<sha256(sessionId) first 32 hex>.orca-mobile-web.invalid/`, so a referrer carries a
    // stable per-session handle rather than the id itself; iOS serves
    // `orca-mobile-web://<sessionId>/` and carries the id verbatim. Either is something
    // an image host can key on across requests, and `img-src https:` is what made it reachable: an
    // image the artifact or a rendered document names is a request to someone else's host. The
    // iframe's own `referrerPolicy` does not cover it -- measured on WebKit, a srcdoc frame's image
    // request carried the embedder's origin anyway, where Chromium sent none -- so the guarantee
    // belongs on the document, where one header covers every request the page makes.
    headers["Referrer-Policy"] = "no-referrer"
  }
  return headers
}

File: mobile/modules/orca-mobile-web-shell/ios/MobileWebShellResponseHeaders.swift
/// The headers one served asset answers with.
///
/// The policy header rides the document and nothing else: on a script or a stylesheet response it
/// is inert, and sending it everywhere would hide which response is the one that has to carry it.
enum MobileWebShellResponseHeaders {
  static func forPath(
    _ path: String,
    contentType: String,
    byteCount: Int
  ) -> [String: String] {
    var headers = [
      "Content-Type": contentType,
      "Content-Length": String(byteCount),
      "Cache-Control": "no-store",
      "X-Content-Type-Options": "nosniff"
    ]
    if path == "/" {
      headers["Content-Security-Policy"] = MobileWebShellCsp.header
      // The document's origin is `orca-mobile-web://<sessionId>/` -- a custom scheme's host is
      // opaque, so the id is used verbatim and a referrer carries it. Android hashes it into a
      // host instead; see its twin. `img-src https:` made that reachable: an image the
      // artifact or a markdown document names is a request to someone else's host. The iframe's
      // own `referrerPolicy` does not cover it -- measured on WebKit, a srcdoc frame's image
      // request carried the embedder's origin anyway, where Chromium sent none -- so the guarantee
      // belongs on the document, where one header covers every request the page makes.
      headers["Referrer-Policy"] = "no-referrer"
    }
    return headers
  }
}
